### PR TITLE
Add support for qThreadExtraInfo

### LINKED ISSUE #104: `qThreadExtraInfo` support
While porting probe-rs to use gdbstub I found two packets that would be helpful to respond to:

~~`qC` - this queries the current thread ID.  WIthout responding GDB works correctly, but the client emits a message that may be confusing to end users.~~  Edit: `qC` is no longer needed after debugging an issue with `?` in the thread.

`qThreadExtraInfo` - this allows the stub to provide a string of extra data about a thread.  It's a free form string that is displayed on the client and can be useful for adding extra detailed status or thread names, for example.

Both of these would produce a better experience for the probe-rs logic and I'd be glad to take a shot at producing a PR that adds the logic.  Before I do I wanted some feedback on how to approach this.  

Both packets feel like appropriate functions on the `MultiThreadBase` trait.  To avoid a break I could add them with default implementations - `qC` could return the first value from `list_active_threads` and `qThreadExtraInfo` could return a 0 byte string by default.  This avoids a break, but it puts a bit more burden on the implementor to realize these are options for them to extend.

Any thoughts / feedback on this?

### PR DESCRIPTION
**Pull request by thefaxman**
### Description

Add a new sub-IDET for `MultiThreadBase` to respond with extra information about a thread.

Closes #104 

### API Stability

- [X] This PR does not require a breaking API change

### Checklist

- Implementation
  - [X] `cargo build` compiles without `errors` or `warnings`
  - [X] `cargo clippy` runs without `errors` or `warnings`
  - [X] `cargo fmt` was run
  - [X] All tests pass
- Documentation
  - [X] rustdoc + approprate inline code comments
  - [X] Updated CHANGELOG.md
  - [X] (if appropriate) Added feature to "Debugging Features" in README.md
- _If implementing a new protocol extension IDET_
  - [X] Included a basic sample implementation in `examples/armv4t`
  - [X] Included output of running `examples/armv4t` with `RUST_LOG=trace` + any relevant GDB output under the "Validation" section below
  - [X] Confirmed that IDET can be optimized away (using `./scripts/test_dead_code_elim.sh` and/or `./example_no_std/check_size.sh`)

### Validation

<details>
<summary>GDB output</summary>

```
(gdb) info thread
  Id   Target Id            Frame 
* 1    Thread 1.1 (CPU Cpu) 0x55550000 in ?? ()
  2    Thread 1.2 (CPU Cop) 0x55550000 in ?? ()
(gdb) 
```

</details>

<details>
<summary>armv4t output</summary>

```
 TRACE gdbstub::protocol::recv_packet     > <-- $qThreadExtraInfo,p1.1#85
 TRACE gdbstub::protocol::response_writer > --> $43505520437075#d2
 TRACE gdbstub::protocol::recv_packet     > <-- $qThreadExtraInfo,p1.2#86
 TRACE gdbstub::protocol::response_writer > --> $43505520436f70#02
```
</details>

### CODE REVIEW

File: CHANGELOG.md
Comment (daniel5151) [inline CHANGELOG.md:5]: you can omit this changelog update. I realized that I usually end up writing these myself anyways, so I'll be removing that step from the PR checklist shortly...

File: examples/armv4t_multicore/gdb.rs
Comment (daniel5151) [inline examples/armv4t_multicore/gdb.rs:295]: I believe this function should already be floating around somewhere in armv4t... no need to redefine it.

File: examples/armv4t_multicore/gdb.rs
Comment (daniel5151) [inline examples/armv4t_multicore/gdb.rs:7]: for consistency, could you use the fully qualified path in the impl block, rather than importing it?

File: src/protocol/commands/_qThreadExtraInfo.rs
Comment (daniel5151) [inline src/protocol/commands/_qThreadExtraInfo.rs:17]: take a look at how “vFilePread` is implemented. 

In this case, since you don't need to hold on to any references into the `buf` after parsing is complete, you are able to reuse the entirety of the buffer.

File: src/stub/core_impl/thread_extra_info.rs
Comment (daniel5151) [inline src/stub/core_impl/thread_extra_info.rs:23]: take a look at how `vFilePread` handles this. `gdbstub` takes care to avoid introducing any panicking paths, which includes those that might occur from array indexing.

File: src/target/ext/thread_extra_info.rs
Comment (daniel5151) [inline src/target/ext/thread_extra_info.rs:19]: nit: `thread_extra_info`?

File: src/stub/core_impl/thread_extra_info.rs
Comment (daniel5151) [inline src/stub/core_impl/thread_extra_info.rs:16]: so, this guard need to be modified to ensure that the rest of the function gets optimized out if `support_thread_extra_info` returns `None`.

The following should do what you want (along with simplifying some of the `ops` code down below as well, removing the need for that `PacketUnexpected` arm).

```suggestion
        let ops = match target.base_ops() {
            BaseOps::SingleThread(_) => return Ok(HandlerStatus::Handled),
            BaseOps::MultiThread(ops) => match ops.support_thread_extra_info() {
                Some(ops) => ops,
                None => return Ok(HandlerStatus::Handled),
            },
        };
```

And please add in the `__dead_code_marker` as present in other handlers, and validate that this bit of code does indeed get dead code eliminated in the case that the target doesn't implement this IDET. (i.e: by temporarily commenting it out in the armv4t example, and running the `test_dead_code_elim.sh` script). 

In addition, please make sure that running `./example_no_std/check_size.sh` returns the same value before / after this PR.

File: src/target/ext/thread_extra_info.rs
Comment (daniel5151) [inline src/target/ext/thread_extra_info.rs:19]: reading through the docs, it seems that `qThreadExtraInfo` is _not_ allowed to return a arbitrary error value. As such, this signature should be modified to return a regular `Result<usize, Self::Error>`.

This is a common inconsistency found in the GDB RSP, where only some methods support reporting error codes, but not others. It's pretty annoying...

**Review by daniel5151 — CHANGES_REQUESTED** [review]
All in all, this is looking good!

File: CHANGELOG.md
Comment (thefaxman) [reply CHANGELOG.md]: Reverted

File: examples/armv4t_multicore/gdb.rs
Comment (thefaxman) [reply examples/armv4t_multicore/gdb.rs]: Done

File: src/protocol/commands/_qThreadExtraInfo.rs
Comment (thefaxman) [reply src/protocol/commands/_qThreadExtraInfo.rs]: Thanks for the pointer, I've matched the logic

File: src/stub/core_impl/thread_extra_info.rs
Comment (thefaxman) [reply src/stub/core_impl/thread_extra_info.rs]: Updated

File: src/target/ext/thread_extra_info.rs
Comment (thefaxman) [reply src/target/ext/thread_extra_info.rs]: Updated

File: src/stub/core_impl/thread_extra_info.rs
Comment (thefaxman) [reply src/stub/core_impl/thread_extra_info.rs]: Updated as requested.

`test_dead_code_elim.sh` shows the path is getting eliminated properly.  `check_size.sh` shows an increase of exactly 1 byte in .text.  I'm still trying to trace down where that is coming from.

File: src/target/ext/thread_extra_info.rs
Comment (thefaxman) [reply src/target/ext/thread_extra_info.rs]: Yep, you're right.  It's clear that the GDB protocol has evolved very organically so there's a lot of pitfalls.

Updated.

File: src/stub/core_impl/thread_extra_info.rs
Comment (daniel5151) [reply src/stub/core_impl/thread_extra_info.rs]: huh, interesting. a single byte difference? that's weird...

well, in any case, I wouldn't worry too much about that 😄

File: examples/armv4t_multicore/gdb.rs
Comment (daniel5151) [reply examples/armv4t_multicore/gdb.rs:294]: ahh, indeed: examples/armv4t/gdb/mod.rs:39

File: src/stub/core_impl/thread_extra_info.rs
Comment (thefaxman) [reply src/stub/core_impl/thread_extra_info.rs]: Looks like the 1 byte difference is just the optimizer reordering some jmp / calls such that we end up with a slightly different generated main code.  None of the calls in question are in the new IDET paths, but it's enough to cause the compiler to generate code slightly differently.

I can provide assembly diffs if interested but my read is that this may be very difficult to avoid.

Comment (daniel5151) [reply src/stub/core_impl/thread_extra_info.rs]: yep, that's what I would've guessed. I wouldn't sweat it - different rustc versions introduce plenty more variation already. the key thing is that we aren't seeing massive bumps in size (i.e: hundreds of bytes).

File: examples/armv4t_multicore/gdb.rs
Comment (thefaxman) [reply examples/armv4t_multicore/gdb.rs:294]: There was a copy in armv4t, but not armv4t_multicore

Comment (daniel5151) [reply examples/armv4t_multicore/gdb.rs:294]: ahh, didn't realize this was `armv4t_multicore`, oops 😅 
All good

File: src/target/ext/thread_extra_info.rs
Comment (thefaxman) [reply src/target/ext/thread_extra_info.rs]: I missed part of this in the second push, should all be fixed to use `Result<usize, Self::Error>` now.

**Comment by thefaxman** [comment]
I think I've addressed all the feedback - let me know if there's any additional items you'd like to see addressed

**Review by bet4it — CHANGES_REQUESTED** [review]
Refer: https://github.com/daniel5151/gdbstub/blob/master/src/protocol/commands/_p.rs

File: src/stub/core_impl/thread_extra_info.rs
Comment (daniel5151) [inline src/stub/core_impl/thread_extra_info.rs:32]: You'll want to add an extra else arm here:

```rust
} else {
    return Err(Error::PacketParse(
        crate::protocol::PacketParseError::MalformedCommand,
    ))
}
```

The _best_ solution would actually be to go the extra mile and introduce a new `ThreadId` variant at the `protocol` level that doesn't support the `All` variant. Something like `ConcreteThreadId` (not a great name, but whatever). That way, an invalid thread-id would be caught at the packet parse level, rather than within the handler.

It would be nice to implement that latter solution, but it's fine if you'd like to stick with the quick-and-dirty fix I proposed above.

File: src/protocol/commands/_qThreadExtraInfo.rs
Comment (thefaxman) [reply src/protocol/commands/_qThreadExtraInfo.rs]: Updated

File: src/protocol/commands/_qThreadExtraInfo.rs
Comment (thefaxman) [reply src/protocol/commands/_qThreadExtraInfo.rs]: Updated

**Review by daniel5151 — CHANGES_REQUESTED** [review]
Looks good.

It's a bit late for me here, so I'll give it one last once-over in the morning just in case, but it should be good to go after resolving this last thing.

File: src/stub/core_impl/thread_extra_info.rs
Comment (thefaxman) [reply src/stub/core_impl/thread_extra_info.rs]: I like that - I added a `ConcreteThreadId` type.  Naming is hard - I couldn't come up with another alternative that I felt expressed things more clearly.

As suggested this simplified the handler logic.

**Comment by daniel5151** [comment]
Thanks again for the contribution! 

Once #103 lands, I'll push out `0.6.2` that includes these and other changes.

**Comment by daniel5151** [comment]
@thefaxman 0.6.2 has been released :)

**Pull request merged**
